### PR TITLE
Rewrite and simplify LED 

### PR DESCRIPTION
**Pull request by edongashi**
After many attempts with various tricks, performance is finally good and reliable.

Removed heuristics and went with an exact approach. The trick was the `n` caches for `n-1` cols. Takes a little longer to build but has much more predictable runtime.

### CODE REVIEW

File: anonymizer/OpenDiffix.Service/Led.fs [hunk 1]
```diff
@@ -43,244 +28,174 @@ let private mergeGivenAggregatorsInto (targetBucket: Bucket) aggIndexes (sourceB
...
+
+  // For each column, we build a cache where we group buckets by their labels EXCLUDING that column.
+  // This means that every cache will associate siblings where the respective column is different.
+  let caches =
+    Array.init groupingLabelsLength (fun _i -> Dictionary<Row, MutableList<Bucket * bool>>(Row.equalityComparer))
+
+  let victimBuckets =
+    buckets
+    |> Array.choose (fun bucket ->
+      let lowCount = Utils.isLowCount lowCountIndex bucket
```
Comment (pdobacz) [inline anonymizer/OpenDiffix.Service/Led.fs:101]: I needed to spend some time grokking this, so let me rephrase what it does, and you tell me if I'm on the track:

We have a single pass through buckets here, but we do two fairly separate things:
1/ choose lowCount buckets to become victimBuckets
2/ populate a "two tier" mapping <column index> -> <bucket values excluding that index> -> <that bucket>. So we effectively cache which buckets would be potential siblings for a given "n-1 element subrows".

If that's more or less the case, maybe there's a way to make it a "two stage but effectively still single pass" like

```
buckets
|> Seq.map(fun which populates caches)
|> Array.choose(fun which chooses lowCount bucket tuples)
```

If those two "tasks" are truly independent, like it seems to me, then this would read easier, and you could provide one-liner comments as to what does each step do, this would make it even easier.

You'd need to carry over the `cacheKeys` from 1st stage to the 2nd, but maybe it's not a big deal?

File: anonymizer/OpenDiffix.Service/Led.fs [hunk 2]
```diff
@@ -43,244 +28,174 @@ let private mergeGivenAggregatorsInto (targetBucket: Bucket) aggIndexes (sourceB
...
+  /// and has merge targets (siblings which match isolating column criteria).
+  let chooseMergeTargets (victimBucket: Bucket, cacheKeys: CacheKey array) =
+    let mutable hasUnknownColumn = false
+    let mergeTargets = MutableList()
+
+    for colIndex in 0 .. groupingLabelsLength - 1 do
+      let cacheKey = cacheKeys.[colIndex]
+      let siblings = caches.[colIndex].[cacheKey]
+
+      if siblings.Count = 1 then
```
Comment (pdobacz) [inline anonymizer/OpenDiffix.Service/Led.fs:142]: so `siblings` includes "self"? I'm not sure I'm following this part. Is this dependent solely on the overall number of columns?

File: anonymizer/OpenDiffix.Service/Led.fs [hunk 3]
```diff
@@ -43,244 +28,174 @@ let private mergeGivenAggregatorsInto (targetBucket: Bucket) aggIndexes (sourceB
...
+        // No siblings for this column (Count=1 means itself)
+        hasUnknownColumn <- true
+
+      if siblings.Count = 2 then
```
Comment (pdobacz) [inline anonymizer/OpenDiffix.Service/Led.fs:146]: hm, how about turning this double `if` into a `match with`, where by we explicitly lay out the implications of how many siblings there were, like so:

```
| 1 -> ...
| 2 -> ...
| 3 -> //do nothing here but explain why
| _ -> failwith "some explanation why this is impossible"
```
?

Asking for this, because I am still unsure I navigate this correctly, the added prose here would probably dispel the doubts

**Review by pdobacz — COMMENTED** [review]
ok, I think I'm getting this, but it takes some time to get through. I left some requests for comments or clarifying refactors, but overall the pieces start to fit together.

**Comment by cristianberneanu** [comment]
> Removed heuristics and went with an exact approach. 

Great 👍 ! As long as the system is usable, this is always preferable. Heuristics make understanding the failure points of the mechanism very painful.

File: anonymizer/OpenDiffix.Service/Led.fs (hunk 2 above)
Comment (edongashi) [reply anonymizer/OpenDiffix.Service/Led.fs:142]: > so siblings includes "self"?

Yes, self will always end up in that slot. I had a comment somewhere for this but must have deleted it while refactoring.

File: anonymizer/OpenDiffix.Service/Led.fs (hunk 1 above)
Comment (edongashi) [reply anonymizer/OpenDiffix.Service/Led.fs:101]: This loop is the most expensive part of the hook. I experimented with parallelization but gains are minimal since there will be contention for dictionary writes (which are not thread safe). Switching to `ConcurrentDictionary` didn't help either.

I agree that the function body is a little dirty with multiple side effects, but I'd favor performance over looks since this is a hot spot.

Comment (pdobacz) [reply anonymizer/OpenDiffix.Service/Led.fs:101]: but would the split as suggested hinder performance? I see that the reason for a single pass is performance, but I think the split would keep the same complexity. Hm maybe the loss would lie with the extra `cacheKeys` passed around.

So nevermind then. I'm out of ideas how to make it easier to grasp :(.

File: anonymizer/OpenDiffix.Service/Led.fs (hunk 3 above)
Comment (edongashi) [reply anonymizer/OpenDiffix.Service/Led.fs]: Yeah, looks better. Fixed.

**Review by pdobacz — APPROVED** [review]

File: anonymizer/OpenDiffix.Service/Led.fs (hunk 1 above)
Comment (edongashi) [reply anonymizer/OpenDiffix.Service/Led.fs:102]: Hah, I am about to make it more complex! I have some ideas with parallelization. If they don't work, I will try to simplify.

Note that there is already a bit of simplification because we no longer return cache keys but just the reference to the (mutable) list of siblings.

**Comment by edongashi** [comment]
Ready for review

**Review by pdobacz — APPROVED** [review]
